routes/cube_routes.js

```diff
@@ -1248,103 +1203,72 @@ router.post('/bulkuploadfile/:id', ensureAuth, async (req, res) => {
   }
 });
 
-async function bulkuploadCSV(req, res, cards, cube) {
-  let added = [];
-  let missing = '';
-  let changelog = '';
-  for (let card_raw of cards) {
-    let split = util.CSVtoArray(card_raw);
-    let name = split[0];
-    let maybeboard = split[8];
-    let card = {
-      name: name,
-      cmc: split[1],
-      type_line: split[2].replace('-', '—'),
-      colors: split[3].split('').filter((c) => [...'WUBRG'].includes(c)),
-      set: split[4].toUpperCase(),
-      addedTmsp: new Date(),
-      collector_number: split[5],
-      status: split[6],
-      finish: split[7],
-      imgUrl: split[9] && split[9] != 'undefined' ? split[9] : null,
-      tags: split[10] && split[10].length > 0 ? split[10].split(',') : [],
-    };
+router.post('/bulkreplacefile/:id', ensureAuth, async function(req, res) {
+  if (!req.files) {
+    req.flash('danger', 'Please attach a file');
+    return res.redirect('/cube/list/' + req.params.id);
+  }
 
-    let potentialIds = carddb.allIds(card);
-    if (potentialIds && potentialIds.length > 0) {
-      // First, try to find the correct set.
-      let matchingSet = potentialIds.find((id) => carddb.cardFromId(id).set.toUpperCase() == card.set);
-      let nonPromo = potentialIds.find(carddb.notPromoOrDigitalId);
-      let first = potentialIds[0];
-      card.cardID = matchingSet || nonPromo || first;
-      if (maybeboard == 'true') {
-        cube.maybe.push(card);
+  const items = req.files.document.data.toString('utf8'); // the uploaded file object
+
+  try {
+    const cube = await Cube.findOne(build_id_query(req.params.id));
+    if (!cube) {
+      req.flash('danger', 'Cube not found');
+      return res.status(404).render('misc/404', {});
+    } else if (cube.owner != req.user._id) {
+      req.flash('danger', 'Not Authorized');
+      return res.redirect('/cube/list/' + req.params.id);
+    }
+    cards = items.match(/[^\r\n]+/g);
+    if (cards) {
+      if (cards[0].trim() == CSV_HEADER) {
```

Comment:
Should we require the header or just remove it if present? Seems reasonable to make it optional, if not then specify in error message that the header is missing?

Reply:
It's the main way we differentiate formats at this point in time. Eventually I'd imagine replacement would support all the other import methods we currently handle like CubeTutor and Plaintext.

Reply:
Ideally we would parse the header and map the column index to the corresponding field, so our csv import would be backwards- compatible

Reply:
I think I'll leave that for a separate PR since this is pretty large as is.